src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs

```diff
@@ -310,33 +331,59 @@ internal override sealed bool TryReadAsObject(ref Utf8JsonReader reader, JsonSer
         }
 
         /// <summary>
-        /// Overridden by the nullable converter to prevent boxing of values by the JIT.
+        /// Performance optimization. Overridden by the nullable converter to prevent boxing of values by the JIT.
+        /// Although the JIT added support to detect this IL pattern (box+isinst+br), it still manifests in TryWrite().
+        /// </summary>
+        internal virtual bool IsNull(in T value) => value is null;
+
+        /// <summary>
+        /// Performance optimization to determine if IsNull() above should be called.
         /// </summary>
-        internal virtual bool IsNull(in T value) => value == null;
+        /// <returns></returns>
+        internal virtual bool IsNullableOfT() => false;
 
+#if NET6_0_OR_GREATER
+        [MethodImpl(MethodImplOptions.AggressiveOptimization)]
+#endif
         internal bool TryWrite(Utf8JsonWriter writer, in T value, JsonSerializerOptions options, ref WriteStack state)
         {
             if (writer.CurrentDepth >= options.EffectiveMaxDepth)
             {
                 ThrowHelper.ThrowJsonException_SerializerCycleDetected(options.EffectiveMaxDepth);
             }
 
-            if (default(T) is null && !HandleNullOnWrite && IsNull(value))
+            // We do not pass null values to converters unless HandleNullOnWrite is true. Null values for properties were
+            // already handled in GetMemberAndWriteJson() so we don't need to check for IgnoreNullValues here.
+            if (_autoWriteNulls)
             {
-                // We do not pass null values to converters unless HandleNullOnWrite is true. Null values for properties were
-                // already handled in GetMemberAndWriteJson() so we don't need to check for IgnoreNullValues here.
-                writer.WriteNullValue();
-                return true;
+                // Optimized check to see if 'IsNull()' or 'is null' should be used.
+                if (_isNullableOfT)
+                {
+                    if (IsNull(value))
```

Comment:
I think we could improve legibility here if all this logic got lifted to a helper method. Here's a sample refactoring that I did on top of main: https://github.com/eiriktsarpalis/runtime/commit/609f635c15a9940d34aafdb96cf325151d8c4c5b which yields a ~8% performance improvement when serializing root-level `DateTimeOffset?` values.

Reply:
Is this still relevant if we just have the LOC `if (default(T) is null && !HandleNullOnWrite && IsNull(value))` where `IsNull` no longer has the `in` modifier?

Reply:
I tried benchmarking that approach and it turned out to be identical to the performance of `main`. I suspect though that there might be performance regressions if `T` happens to be a large nullable struct.

Reply:
>  identical to the performance of main

Do you mean without the `in` modifier in `IsNull`?

With the `in` modifier, the tests that use `Nullable<>` alloc\box and are much slower.

Reply:
Compared to what we have in main currently, i.e In modifier in a virtual method.

Reply:
So eiriktsarpalis@609f635 provides an 8% improvement on top of that?

Reply:
Correct. Might make sense to measure vs the current head of the PR branch.

Reply:
Sure, I'll give it a try. If the results are mixed, I'm inclined to favor the common object\valuetype\nullable scenarios instead of the large-value-type (anti-pattern) POCOs, meaning favoring `if (default(T) is null && !HandleNullOnWrite && IsNull(value))` instead.

Reply:
I tried rebasing https://github.com/eiriktsarpalis/runtime/commit/609f635 over your PR branch but the benchmark results were identical. I think we can keep as-is.

Reply:
Thanks for test that.